### PR TITLE
Support kubectl debug overload from overrides json

### LINKED ISSUE #97103: [enhancement] kubectl alpha debug add SYS_PTRACE for debug container
```
VERSION: v1.18.3
Command:  kubectl alpha debug -it xxx-854d568b99-klgc9 --image=myimage:latest --container=xxx-854d568b99-klgc9 --target=xxx
In Debug Container: I use PTRACE for my process , but I get error message: 
Failed to setns 4: Operation not permitted
```
**What would you like to be added**:
I hope that this debug feature can add "SYS_PTRACE" for container

**Why is this needed**:
We always to analysis  our process in POD , this PTRACE is very useful

### PR DESCRIPTION
**Pull request by wawa0210**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #97103

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubectl debug support to specify capabilities
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
kubectl debug support to specify capabilities
```

/sig cli

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:180]: nit: s/,set/, set/

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:602]: Please remove the debugging messages, here and below.

**Comment by liggitt** [comment]
Is this command expected to grow flags for every container field? in the past we've decided against this approach (e.g. in `kubectl create $type` since specifying structured information via flags quickly gets unwieldy. Rather than a piecemill approach like this, would it make sense to allow passing a merge patch that gets merged with the container spec we expect to send?

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:602]: drop debug logging

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:605]: drop debug logging

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:607]: this stomps any existing security context (and I didn't think security contexts could be set in ephemeral containers currently)

Comment (verb) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:607]: That's correct, though this method isn't used with ephemeral containers and is modifying a fresh copy of the pod.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:603]: This should be done in `Complete()`

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:114]: Store capabilities here in options. You can do the conversion in `Complete()`.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (wawa0210) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:86]: thx,done

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (wawa0210) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go]: Thanks, completed

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (wawa0210) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go]: Thanks, completed

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:180]: Use `StringSlice()` here instead.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:238]: make this a single statement, like:

```
if capStrings := cmdutil.GetFlagString(...); len(capStrings) > 0 {
```

**Review by verb — CHANGES_REQUESTED** [review]
Before going any further we need to resolve the UX. Do we want to have users specifying a list a capabilities via flag? I haven't received any suggestions from my post to the sig-cli mailing list.

If we don't work it out before then, I'll bring it up at the sig-cli meetup next Wednesday. FYI @soltysh

**Comment by wawa0210** [comment]
> If we don't work it out before then, I'll bring it up at the sig-cli meetup next Wednesday. FYI @soltysh

Does sig-cli have the latest feedback on this pr? Does the community accept the current plan?

**Comment by verb** [comment]
@wawa0210 I have it on the agenda for tonights sig-cli meetup

**Comment by verb** [comment]
> @wawa0210 I have it on the agenda for tonights sig-cli meetup

We discussed some ways of giving a better user experience over the long run but agreed that a good first step is to provide a flag similar to `kubectl run --override` that allows specifying a patch as a string or file.

**Comment by wawa0210** [comment]
> We discussed some ways of giving a better user experience over the long run but agreed that a good first step is to provide a flag similar to `kubectl run --override` that allows specifying a patch as a string or file.

I go to learn about the implementation of `kubectl run --override`, and then try to mention a pr, but it may take some time

**Comment by verb** [comment]
Sounds good @wawa0210, thanks for working on this.

/hold

**Comment by wawa0210** [comment]
> Sounds good @wawa0210, thanks for working on this.
> 

I researched it, not sure whether the process is completely correct, the conclusion is as follows:
It is not operable to specify parameters through a similar `kubectl run --override`

The debug command is `kubectl debug node/nodeName -it --image=busybox`,the debugpod yaml generated by this command is(only show continer key information)
```
{
	"spec": {
		"containers": [{
			"name": "debugger",
			"image": "10.7.106.189/daocloud/busybox",
			"resources": {},
			"volumeMounts": [{
				"name": "host-root",
				"mountPath": "/host"
			}, {
				"name": "default-token-lwjgm",
				"readOnly": true,
				"mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
			}],
			"terminationMessagePath": "/dev/termination-log",
			"terminationMessagePolicy": "File",
			"imagePullPolicy": "Always",
			"stdin": true,
			"tty": true
		}]
	}
}
```

If we override debugPod yaml through `overrides`
My test yaml is `{"spec":{"containers":[{"securityContext":{"capabilities":{"add":["NET_ADMIN","SYS_TIME"]}}}]}}`

But `jsonpatch.MergePatch` cannot merge as we expected, the actual result is

```
{
	"spec": {
		"containers": [{
			"securityContext": {
				"capabilities": {
					"add": ["NET_ADMIN", "SYS_TIME"]
				}
			}
		}]
	}
}
```

The merge of the array cannot be completed well, and the configuration other than the debugPod securityContext has been overwritten, which does not match expectations.

At the same time, many configurations of the debugPod container are dynamically obtained at runtime, such as `volumeMounts`,`container-name`. The user does not know these values ​​when using `--overrides`

Not sure if my test steps are completely correct

**Comment by mradmila** [comment]
What's the current status on this PR? It would be really nice to see this one finally in master.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:86]: I share the same concern as Jordan, I don't want `kubectl debug` to end up just like `kubectl run` with too many flags which only complicate their use. For now I'd like to hear how many use-cases are for such granular flags. I'm inclined to do something like `--overrides` in `kubectl run` for those cases rather than adding more and more flags. 
/hold

**Review by soltysh — CHANGES_REQUESTED** [review]

**Comment by mradmila** [comment]
I completely understand your concern, and Jordan's. However, I am approaching this from a pragmatic side. kubectl debug is completely useless to me without SYS_PTRACE capability, so any solution that will finally allow me to properly debug distroless containers will be fine. Whether this is specified through overrides, or special flags I will leave to you and other maintainers.

**Comment by wawa0210** [comment]
@liggitt @verb @soltysh

**Comment by wawa0210** [comment]
@liggitt

frindly ping, can you help review this pr again, thx

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:627]: This is not right, we should be laveraging our internal codec mechanism, compare with https://github.com/kubernetes/kubernetes/blob/19c7089245268e59e46b0a19da9fed45d18cd73f/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L624-L630 which has that same functionality.

**Review by soltysh — CHANGES_REQUESTED** [review]

**Comment by verb** [comment]
> 
> At the same time, many configurations of the debugPod container are dynamically obtained at runtime, such as `volumeMounts`,`container-name`. The user does not know these values ​​when using `--overrides`

@wawa0210 The user doesn't know these necessarily, but `kubectl` does, right? Can't we add the correct container name before generating the patch?

**Pull request closed without merging**
